### ISSUE TITLE
experimentalPreserveModules leaves in unused import statements

### ISSUE THREAD

**Issue by matthewp** [rollup/rollup#2120]
Here's a sample repo showing the problem. https://github.com/matthewp/rollup-bug

In short we have:

__main.js__

```js
import { a } from './one';

window.APP = { a };
```

__one.js__

```js
export { default as a } from './two';
export { default as b } from './three';
```

As you can see, the __three.js__ module is not used. This gets transformed into:

__main.js__

```js
import './one.js';
import a from './two.js';

window.APP = { a };
```

__one.js__

```js
import './two.js';
import './three.js';
```

As you can see, the __two.js__ module, which is used, gets promoted up to being an import for __main.js__. But __one.js__ still includes the import for __three.js__ even though it is not used in the application at all.

I understand that it might work this way because rollup cleans this up when it bundles. If you are using `experimentalPreserveModules`, however, you might be using something else for bundling (or perhaps not bundling at all), in which case it creates an unnecessary request for __three.js__.

**Comment by matthewp** [comment]
I would think that when using `experimentalPreserveModules` it should not add these side-effectual import statements (the `import 'foo'` variety) at all, unless they were originally written that way.

Would you be open to a pull request which updates this?

**Comment by matthewp** [comment]
@guybedford Thanks, I'll try that out and report back.

**Comment by lukastaegert** [comment]
If I looked correctly it is actually an internal import that is causing problems. I believe this is a known issue for preserve modules that at the moment it is not able to identify tree-shaken modules (even though the statements themselves are removed). Once #2119 lands, we will have an easy way of checking if a module was entirely tree-shaken which we could use to both prevent emitting the file and prevent the empty import (since #2119 will put an .included flag on Program nodes)

**Comment by matthewp** [comment]
`treeshake: { pureExternalModules: true }` did not work. @lukastaegert If I understand you correctly I think that will work.

**Comment by guybedford** [comment]
There are actually two separate issues here:

1. The pureExternalModules bug is happening in that `one.js` should emit `import './two.js'` only.
2. As @lukastaegert describes, once we have `included` on modules themselves we can strip the `one.js` empty file entirely from the build.

If we can aim to fix (1) now, then we'll be setup to handle (2).

I think I may actually know the issue here in (1), will see if I can look into it, although PR for a test case or fix welcome too.

**Comment by guybedford** [comment]
Note that on master `main.js` contains `import 'three.js'` which is actually a third bug :)

**Comment by matthewp** [comment]
I can open a separate issue to discuss if you all prefer, but related is that I don't think there will be a way to know what dependencies a module has (or conversely which dependencies were pruned) without reparsing the module and comparing to what's in `bundle`. Can metadata about dependencies post-generate be added somewhere?

**Comment by matthewp** [comment]
Looking more into it, I'm not sure why `pureExternalModules` is part of this bug. __three.js__ is not an external module...

**Comment by guybedford** [comment]
Ah, right, I misread three.js as an external, yes this is just about
module-level pruning then as Lukas mentioned.

**Comment by matthewp** [comment]
I'm added a test case in https://github.com/stealjs/rollup/commit/83d3f841ba62965bd8f9051ddbe9e0cfe1f124f5
